keps/sig-node/20200110-apparmor-ga.md

```diff
@@ -0,0 +1,387 @@
+to be added in the future (e.g. Kubernetes predefined profiles or ConfigMap
+profiles). The AppArmor options structure leaves room for future extensions,
+such as defining the behavior when a profile cannot be set.
+
+#### PodSecurityPolicy API
+
+```go
+type PodSecurityPolicySpec struct {
+    ...
+    // AppArmor is the strategy that will dictate allowable and default AppArmor
+    // profiles for the container.
+    // +optional
+    AppArmor *ApparmorStrategyOptions
+    ...
+}
+
+type AppArmorStrategyOptions struct {
+    // The default profile to set on the pod, if none is specified.
+    // The default MUST be allowed by the allowedProfiles.
+    // +optional
+    DefaultProfile *v1.AppArmorProfile
+
+    // The set of profiles that may be set on the pod or containers.
+    // If unspecified, AppArmor profiles are unrestricted by this policy.
+    // +optional
+    AllowedProfiles *AppArmorProfileSet
+}
+
+// A set of AppArmor profiles. This struct should be a plural of
+// `v1.AppArmorProfile`.
+// All values are optional, and an unspecified field excludes all profiles of
+// that type from the set.
+type AppArmorProfileSet struct {
+    // The allowed AppArmor profile types.
+    // +optional
+    Types []AppArmorProfileType
+
+    // The allowed runtimeProfiles. A value of '*' allows all runtimeProfiles.
+    // +optional
+    RuntimeProfiles []string
+
+    // The allowed localhostProfiles. Values may end in '*' to include all
+    // localhostProfiles with a prefix.
+    // +optional
+    LocalhostProfiles []string
+}
+```
+
+## Design Details
+
+### Version Skew Strategy
+
+All API skew is resolved in the API server. New Kubelets will only use the
+AppArmor values specified in the fields and ignore the annotations therefore.
+
+#### Pod Creation
+
+If no AppArmor annotations or fields are specified, no action is necessary.
+
+If _only_ AppArmor fields are specified, add the corresponding annotations. This
+ensures that the fields are enforced even if the node version trails the API
+version. Since [we
+support](https://kubernetes.io/docs/setup/release/version-skew-policy) up to 2
+minor releases of version skew between the master and node, this behavior will
+be removed in version N+4 (where N is the release with the upgraded AppArmor
+support).
+
+If _only_ AppArmor annotations are specified, copy the values into the
+corresponding fields. This ensures that existing applications continue to
+enforce AppArmor, and prevents the kubelet from needing to resolve annotations &
+fields.
+
+If both AppArmor annotations _and_ fields are specified, the values MUST match.
+This will be enforced in API validation.
+
+To raise awareness of annotation usage (in case of old automation), an
+additional warning annotation will be added when a pod is created with an
+AppArmor annotation:
+
+```
+warning.kubernetes.io/apparmor: "AppArmor set through annotations. Support will be dropped in v1.22"
+```
```

Comment:
For seccomp we removed this specific approach and left a more generic entry, mentioning that a warning mechanism would be used, then highlighting which ones were being considered:

> The mechanisms being considerated are audit annotations, annotations on the object, events, or a 
> warning as described in [KEP #1693](/keps/sig-api-machinery/1693-warnings).

Reply:
Thanks, changed that part as well.